### PR TITLE
Strategic bombing bug fix attempt #2

### PR DESCRIPTION
**Pull request by simon33-2**
New PR replacing #1074 

Fix implements property which enables inbuilt AA to fire at only bombers attacking that are attacking that unit, which is needed for Global 1940.

@DanVanAtta @RoiEXLab

---
This change is [Reviewable](https://reviewable.io/reviews/triplea-game/triplea/1133)

### CODE REVIEW

**Comment by RoiEXLab** [comment]
I worked \:D

**Comment by simon33-2** [comment]
Still got the issue that the new test doesn't work. Comment out the "assert" lines from the program?

**Comment by simon33-2** [comment]
Ok, that's done.

File: test/games/strategy/triplea/delegate/GameDataTestUtil.java [hunk 1]
```diff
@@ -80,6 +80,27 @@ public static UnitType carrier(final GameData data) {
     return unitType(Constants.UNIT_TYPE_CARRIER, data);
   }
 
+  // Some units hard coded here rather than placed in Constants at the insistence of community members Ref: Pull Request 1074
```
Comment (DanVanAtta) [inline test/games/strategy/triplea/delegate/GameDataTestUtil.java]: Just a bit passive aggressive?
This is otherwise a very distracting comment and is unnecessary.

File: test/games/strategy/triplea/xml/LoadGameUtil.java [hunk 2]
```diff
@@ -18,7 +18,7 @@
     BIG_WORLD_1942("big_world_1942_test.xml"), IRON_BLITZ("iron_blitz_test.xml"), LHTR(
         "lhtr_test.xml"), PACIFIC_INCOMPLETE("pacific_incomplete_test.xml"), PACT_OF_STEEL_2(
             "pact_of_steel_2_test.xml"), REVISED("revised_test.xml"), VICTORY_TEST("victory_test.xml"), WW2PAC40(
-                "ww2pac40_test.xml"), WW2V3_1941("ww2v3_1941_test.xml"), WW2V3_1942("ww2v3_1942_test.xml");
+                "ww2pac40_test.xml"), WW2V3_1941("ww2v3_1941_test.xml"), WW2V3_1942("ww2v3_1942_test.xml"), Global1940("ww2_g40_balanced.xml");
```
Comment (DanVanAtta) [inline test/games/strategy/triplea/xml/LoadGameUtil.java]: constants get caps, ie: `GLOBAL_1940`

**Comment by simon33-2** [comment]
Updated to resolve issues referred to above.

File: src/games/strategy/triplea/delegate/StrategicBombingRaidBattle.java [hunk 3]
```diff
@@ -203,7 +203,19 @@ public void fight(final IDelegateBridge bridge) {
     showBattle(bridge);
     final List<IExecutable> steps = new ArrayList<>();
     if (hasAA) {
-      steps.add(new FireAA());
+      boolean foundSpecificAA = false;
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/delegate/StrategicBombingRaidBattle.java]: You can tighten this block up with:
```    
if (hasAA) {
      // global1940 rules - each target type fires an AA shot against the planes bombing it
      m_targets.entrySet().stream()
          .filter(entry -> entry.getKey().getUnitAttachment().getIsAAforBombingThisUnitOnly())
          .forEach(entry -> steps.add(new FireAA(entry.getValue())));

      // otherwise fire an AA shot at all the planes
      if(steps.isEmpty()) {
        steps.add(new FireAA());
      }
    }
```
The above is NPE safe without the NPE checks, and avoids the tracking variable `foundSpecificAA`

File: test/games/strategy/triplea/delegate/StratBombTest.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+          break;
+      }
+    }
+    targets.put(airfield, new HashSet<>(Collections.singleton(tacBomber1)));
+    targets.put(harbour, new HashSet<>(Collections.singleton(tacBomber2)));
+    targets.put(factory, new HashSet<>(Collections.singleton(stratBomber)));
+    final ITestDelegateBridge bridge = getDelegateBridge(british);
+    tracker.addBattle(new RouteScripted(germany), attackers, true, british, bridge, null, null, targets, true);
+                
+    final StrategicBombingRaidBattle battle = new StrategicBombingRaidBattle(germany, m_data, british, tracker);
```
Comment (DanVanAtta) [inline test/games/strategy/triplea/delegate/StratBombTest.java]: BattleTracker creates a battle, that is the one that is active. The battle created on this line is extra. 

For a rough version:
https://gist.github.com/DanVanAtta/83d76173f76dcb868fe2b789eaca9a6a

This line is the key:
` final StrategicBombingRaidBattle battle = (StrategicBombingRaidBattle) tracker.getPendingBattles().iterator().next();`

File: test/games/strategy/triplea/delegate/StratBombTest.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    TripleAUnit airfield = null;
+    TripleAUnit harbour = null;
+    TripleAUnit factory = null;
+
+    // Find facilities in territory
+    for( final Unit target : germany.getUnits().getUnits() ) {
+      switch( target.getType().getName() ) {
+        case "airfield"      : airfield = (TripleAUnit) target; 
+          break;
+        case "harbour"       : harbour = (TripleAUnit) target;
```
Comment (DanVanAtta) [inline test/games/strategy/triplea/delegate/StratBombTest.java]: If the AA fire is broken up between harbour and factory, what does throwing in airfield show? IE: in what cases would we see a test failure only because we included airfield in the test data?

File: test/games/strategy/triplea/delegate/StratBombTest.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    battle.fight(bridge);
+    final int airfieldDmg = airfield.getUnitDamage();
+    final int harbourDmg = harbour.getUnitDamage();
+    final int factoryDmg = factory.getUnitDamage();
+    System.out.format("airf: %d harb: %d fact: %d\n", airfieldDmg, harbourDmg, factoryDmg );
+    // targets dice is 4, so damage is 1 + 4 = 5
+    // bomber 2 hits at 2, so damage is 3, for a total of 8
+    // Changed to match StrategicBombingRaidBattle changes
+    // All tests fail. Remove to get past automated testing. Change works in normal game. Not sure why it doesn't work here. Probably some problem with the test.
+    //assertEquals(0, airfieldDmg);
```
Comment (DanVanAtta) [inline test/games/strategy/triplea/delegate/StratBombTest.java]: Would it be more direct to instead verify that the expected bomber was shot down? IE: the unit should be marked as destroyed. It is only indirect evidence here because of missing bomber damage.

File: test/games/strategy/triplea/delegate/StratBombTest.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    m_data = LoadGameUtil.loadTestGame(LoadGameUtil.TestMapXml.GLOBAL1940);
+  }
+
+  private ITestDelegateBridge getDelegateBridge(final PlayerID player) {
+    return GameDataTestUtil.getDelegateBridge(player, m_data);
+  }
+
+
+  @Test
+  public void TestBombingRaid2targets() {
```
Comment (DanVanAtta) [inline test/games/strategy/triplea/delegate/StratBombTest.java]: I'd advocate that you test a minimum to show what you need - we have almost identical tests elsewhere and there little point to duplicate that test coverage (it is really not a good thing, you wind up in the state where one change breaks all tests, and not exactly one).

Regardless, the name of the method is a bit off:
- there are 3 targets here, not 2.
- we can describe that we are testing global allocation, something like `buildingsShootAtOnlyTheirBombers`, the method name that we are testing bombing targets is misleading since we are actually testing AA fire allocation.

File: test/games/strategy/triplea/delegate/StratBombTest.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+import games.strategy.triplea.attachments.UnitAttachment;
+import games.strategy.triplea.player.ITripleAPlayer;
+import games.strategy.triplea.xml.LoadGameUtil;
+
+public class StratBombTest {
+  private GameData m_data;
+
+  @Before
+  public void setUp() throws Exception {
+    m_data = LoadGameUtil.loadTestGame(LoadGameUtil.TestMapXml.GLOBAL1940);
```
Comment (DanVanAtta) [inline test/games/strategy/triplea/delegate/StratBombTest.java]: A comment here stating we are using global1940 to pick up rules where structures fire at only their bombers would be good context to capture. When we look at this code again, we will not wonder why global 1940 was chosen and not another test map.

File: test/games/strategy/triplea/delegate/StratBombTest.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+                
+    final StrategicBombingRaidBattle battle = new StrategicBombingRaidBattle(germany, m_data, british, tracker);
+    battle.addAttackChange(m_data.getMap().getRoute(uk, germany), uk.getUnits().getMatches(Matches.UnitIsStrategicBomber), null);
+    addTo(germany, uk.getUnits().getMatches(Matches.UnitIsStrategicBomber));
+    tracker.getBattleRecords().addBattle(british, battle.getBattleID(), germany, battle.getBattleType());
+    // aa guns rolls 1,3,2 first one hits, remaining bombers roll 1 dice each
+    bridge.setRandomSource(new ScriptedRandomSource(new int[] {1, 3, 2, 5, 4}));
+    // if we try to move aa, then the game will ask us if we want to move
+    // fail if we are called
+    final InvocationHandler handler = new InvocationHandler() {
```
Comment (DanVanAtta) [inline test/games/strategy/triplea/delegate/StratBombTest.java]: Please remove this, there is no failure here, so it is not called. We already test this in other tests. KISS, this test can be made simpler without losing test coverage.

Comment (DanVanAtta) [inline test/games/strategy/triplea/delegate/StratBombTest.java]: Unless it is called, in which case the comment "  // fail if we are called" is a lie.

**Comment by DanVanAtta** [comment]
Glad to see you were able to re-apply the updates.

- Left some comments that will help make future understanding and maintenance easier. We're getting close on this.
- Found the problem with the test (intelliJ has a strong debugger, eclipse would have done same though), see the comments. Points for including a new test case : )
- Once the test is 'fixed' though, the damage applied to the buildings was more than I expected, might need some further investigation.
- Last, and a nitpick, please run the code formatter.

Looks like this is the first fix for the global rules that have a NA on them, really good to see this get picked up.

File: test/games/strategy/triplea/delegate/GameDataTestUtil.java [hunk 10]
```diff
@@ -80,6 +80,27 @@ public static UnitType carrier(final GameData data) {
     return unitType(Constants.UNIT_TYPE_CARRIER, data);
   }
 
+  // Some units hard coded here rather than placed in Constants at the request of community members Ref: Pull Request 1074
```
Comment (DanVanAtta) [inline test/games/strategy/triplea/delegate/GameDataTestUtil.java]: Appreciate the change. I still think it is distracting. This is a good item for us to discuss here though. Why are you of the opinion that these really must be defined in a Constants class. As is, we have some benefits:
- the values are not shared with anything else, so constant does not help remove duplication
- we have better cohesion with the test data being in the test source
- moving to constants would create a layer of indirection, not sure how that is useful since the test data is static.
- These constants are used in test only code for now. If it were in Constants.java, then the assumption would be that we use the values in production code.

File: test/games/strategy/triplea/delegate/StratBombTest.java (hunk 5 above)
Comment (simon33-2) [inline test/games/strategy/triplea/delegate/StratBombTest.java]: The difference is that the AA for the airfield is supposed to hit and leave it undamaged.

File: src/games/strategy/triplea/delegate/StrategicBombingRaidBattle.java (hunk 3 above)
Comment (simon33-2) [inline src/games/strategy/triplea/delegate/StrategicBombingRaidBattle.java]: NPE?

If you say so.

File: test/games/strategy/triplea/delegate/GameDataTestUtil.java (hunk 10 above)
Comment (simon33-2) [inline test/games/strategy/triplea/delegate/GameDataTestUtil.java]: I would have thought it would be consistent with the rest of the system for it to be defined there.

While we're discussing these things, why on earth would anyone restrict to 80 columns in a GUI world? That is a throw back to when code had to be written for an 80x24 Wyse terminal. It makes the code harder to follow and you can fit less on the screen by doing it. Better to have the whole statement on a single line where at all practicable.

File: src/games/strategy/triplea/delegate/StrategicBombingRaidBattle.java (hunk 3 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/delegate/StrategicBombingRaidBattle.java]: null pointer exception

File: test/games/strategy/triplea/delegate/StratBombTest.java (hunk 4 above)
Comment (simon33-2) [inline test/games/strategy/triplea/delegate/StratBombTest.java]: Ah, maybe that's the problem then. I'll try it.

File: test/games/strategy/triplea/delegate/StratBombTest.java (hunk 5 above)
Comment (DanVanAtta) [inline test/games/strategy/triplea/delegate/StratBombTest.java]: Yes, but why do you need two other structures to show this? Is it not sufficient to show that the airfield was left untouched and the other building was damaged?

File: test/games/strategy/triplea/delegate/GameDataTestUtil.java (hunk 10 above)
Comment (DanVanAtta) [inline test/games/strategy/triplea/delegate/GameDataTestUtil.java]: > "I would have thought it would be consistent with the rest of the system for it to be defined there."

I think you raise a good point. But, the pattern is not very good since we have test only code in production code. Consistency to a bad pattern sometimes can be better than inconsistency, in this case the inconsistency has a number of benefits (most significant being the split of test/prod code) 

--- -
RE: line length
120 is the limit per Veqryn. Until not long ago (9 months ago) I actually had a screen that only rendered 90 characters wide comfortably (was a very nice upgrade to a 2k monitor) . 
One benefit is reducing horizontal scrolling. Code is often viewed in console windows, or web browsers, or side by side, and in many contexts really long lines cannot be seen fully. 

It is also to encourage lines that are not super long. Part of the trick is that the indentation should not be very deep, only about 6 or 8 characters from the left, this gives 72-74 characters, or in our case 112 to 114 characters per line - which is a good bit of code for any one line.

File: test/games/strategy/triplea/delegate/StratBombTest.java (hunk 6 above)
Comment (simon33-2) [inline test/games/strategy/triplea/delegate/StratBombTest.java]: Probably, but either approach will work.

As for whether we need 2 tac bombers; perhaps not although I feel that the test is stronger with them both. In fact, it would be better if we shot down the one attacking the harbour because currently if you select the tac, the one attacking the airfield is removed anyway.

**Comment by simon33-2** [comment]
Hmm, I've updated the code along the lines you've suggested but it seems the "addTo" function now causes it to fall over. Here's an extract:

    tracker.addBattle(new RouteScripted(wgermany), attackers, true, british, bridge, null, null, targets, true);
                
    final StrategicBombingRaidBattle battle = (StrategicBombingRaidBattle) tracker.getPendingBattle(wgermany, true, null);
    battle.addAttackChange(m_data.getMap().getRoute(uk, wgermany), uk.getUnits().getMatches(Matches.UnitIsStrategicBomber), null);
    addTo(wgermany, uk.getUnits().getMatches(Matches.UnitIsStrategicBomber));
    tracker.getBattleRecords().addBattle(british, battle.getBattleID(), wgermany, battle.getBattleType());
    // aa guns rolls 1,3,2 first one hits, remaining bombers roll 1 dice each


By removing the addTo, the code runs but doesn't actually do anything still. As it is, it falls over with:
games.strategy.triplea.delegate.StratBombTest > TestBombingRaidInvidualAA FAILED
    java.lang.IllegalStateException: Duplicate Units Detected: Original List:[bomber owned by British, bomber owned by British, bomber owned by British, tactical_bomber owned by British, tactical_bomber owned by British, tactical_bomber owned by British, tactical_bomber owned by British]  HashSet:[tactical_bomber owned by British, bomber owned by British, tactical_bomber owned by British, bomber owned by British]

File: test/games/strategy/triplea/delegate/GameDataTestUtil.java (hunk 10 above)
Comment (simon33-2) [inline test/games/strategy/triplea/delegate/GameDataTestUtil.java]: You must have been using some really old monitor then. Like 10 years old. 1920x1280 capable monitors have been around for a pretty long time.

Console windows? You mean like the Wyse terminal I was referring to above? I doubt much Java is worked on that way and I think it adds strongly negative value to cater to it.

Comment (DanVanAtta) [inline test/games/strategy/triplea/delegate/GameDataTestUtil.java]: ie; linux terminals. I do git work on them, and do a number of diffs there - so the  use-case is there. Google did a good bit of research into the guidelines, though found it was a small but incrementally very significant benefit to not require horizontal scrolling.

RE: monitor, it was quite old, but also keep in mind there were bars on the sides for package structure and outline, so it's not as if the full screen real estate was horizontally available. But, case in point, github only shows so many characters on the website.

File: test/games/strategy/triplea/delegate/StratBombTest.java (hunk 7 above)
Comment (DanVanAtta) [inline test/games/strategy/triplea/delegate/StratBombTest.java]: You can save the effort of 'yes,ok' by either giving a thumbs up reaction, means you agree, or just push a change that fixes this. Ultimately the preference is for us both to spend time on the code, and not on CR overhead

**Comment by DanVanAtta** [comment]
I remember running into that, notice in the gist the addTo is commented out: https://gist.github.com/DanVanAtta/83d76173f76dcb868fe2b789eaca9a6a#file-gistfile1-txt-L45

W.R.T to two tactical bombers vs 1. In this case I would actually advocate most for having two different test cases. One that tests the bare minimum. Then, parameterize the first test, and write a second one that sends additional test data cases. 

A pretty good principle of testing is to add no more test code than is necessary to demonstrate a failure. But, going back to the testing pyramid (https://watirmelon.files.wordpress.com/2011/06/automatedtestingpyramid.png), this is not a unit test, we should try to reduce how many of these kinds of tests we write in favor of more unit tests.

Regardless, let's focus on getting the test case working well, there is not much outstanding feedback. Thank you for your work on this.

**Comment by simon33-2** [comment]
Ok, I have to go out for a while. I've pushed up the changes I have so you can see them.

**Comment by DanVanAtta** [comment]
Bad timing, I broke the build for about 20 minutes. Rebase to latest and there should be a build fix in it.

**Comment by DanVanAtta** [comment]
Merging now so we can get this version deployed as the 1.9 lobby. (this update has been manually tested, I've also checked that the latest master has build fixes).

@simon33-2 Pls  run a `git checkout -b <new branch>` to start on a fresh branch, rebase it and continue work there.

**Comment by simon33-2** [comment]
Awesome. Thanks. 

I can't see the like button BTW.

Happy to help in whatever way I can with the testing, but without much knowledge of the code it is difficult to understand why it isn't working. 

Next thing I want to get sorted is the automated combat selection idea. Can you explain a bit more your idea to avoid sub classing?
